### ISSUE TITLE
Remix.js dev server causing `node:worker_thread` silent error

### ISSUE THREAD

**Issue by AjaniBilby** [remix-run/remix#8155]
### Reproduction

[Minimal Error Reproduction Repo](https://github.com/AjaniBilby/remix-worker-error)

## Dev
```
npm i
npm run dev
```
Goto [`http://localhost:3000/status`](http://localhost:3000/status), should see `PDF Worker: TIMEOUT` - or worst case it will just never load, also based on the logs in `~/pdf/interface.server.ts` the worker is clearly successfully initalized, plus the startup handshake test successfully passes.

It is only after leaving the server idle for a couple of micro-seconds that suddenly the worker_therad no longer responds, and often the entire dev server itself stops responding completely.

However none of this happens if you build the server fully, then run the fully built project.

## Production
```
npm i
npm run build
npm run server
```
Goto [`http://localhost:3000/status`](http://localhost:3000/status), should see `PDF Worker: OK`

You can also go to [`http://localhost:3000/pdf`](http://localhost:3000/pdf) which will show a pdf successfully being rendered in the worker and sent as a response to the client.

### System Info

```shell
Microsoft Windows [Version 10.0.22621.2715]
npm 10.1.0
node v20.7.0
remix v2.3.1
```
```


### Used Package Manager

npm

### Expected Behavior

The dev server and production server should be have almost identically with the exception of hot reloads having the possiblity of breaking things.

There are no error messages at all, and even attempting to use `global.` to ensure the worker persists between hot-reloads does not help the issue.
I would understand if the worker breaks during hot-reloads, but the server will self combust when no changes have taken place since inital start.

### Actual Behavior

No error messages, no variables becoming unbound due to reloading, the server sometimes completely stalls (refuses to respond) without any indication

**Comment by AjaniBilby** [comment]
Since this is quite a complicated issue to do with [`worker_threads`](https://nodejs.org/api/worker_threads.html) and [`comlink`](https://www.npmjs.com/package/comlink) - I don't expect someone to go out of their way to diagnose this directly.

I'm just wondering if there are any tips from the remix devs that might help me find the actual root cause so I can report it - because there is something really weird happening here.

**Comment by AjaniBilby** [comment]
I am also following the recommended singleton pattern from [the docs](https://remix.run/docs/en/main/guides/manual-mode#keeping-in-memory-server-state-across-rebuilds) to keep my worker instance across hot-reloads, but it makes almost no difference.

**Comment by hi-ogawa** [comment]
I tested your repo on my PC (linux) and I can reproduce the same behavior, but I couldn't tell what's the cause.
Ideally, I think your repro could be more minimal. For example, are `@react-pdf/renderer`,  `comlink`, custom `server.ts` etc... necessary to trigger the issue? The natural steps for further debugging would be to remove potential culprit one by one.

Just out of curiosity, I was attempting to make simpler repro in https://github.com/hi-ogawa/remix-dev-node-worker-thread-repro, but I couldn't reproduce the issue. I tried following scenarios:
- `node:worker_threads` (esm) - [main](https://github.com/hi-ogawa/remix-dev-node-worker-thread-repro)
- `node:worker_threads` (cjs) - [chore-cjs](https://github.com/hi-ogawa/remix-dev-node-worker-thread-repro/tree/chore-cjs)
- `node:worker_threads` + `comlink` (esm) - [test-comlink](https://github.com/hi-ogawa/remix-dev-node-worker-thread-repro/tree/test-comlink)
- `node:worker_threads` + `comlink` (cjs) - [test-comlink-cjs](https://github.com/hi-ogawa/remix-dev-node-worker-thread-repro/tree/test-comlink-cjs)

One quick debugging I did on your repro was to check these worker events:

```tsx
// in app/pdf/interface.server.ts
  for (const event of ["error", "exit", "message", "messageerror", "online"]) {
    worker.on(event, (e) => {
      console.log("[worker-client]", event, e);
    })
  }
```

and I saw following logs during dev, which indicates the worker exited somehow:

```
[worker-client] online undefined
PDF Worker initialized
Running Worker Setup
[worker-client] exit 1
```

**Comment by AjaniBilby** [comment]
I made a new branch called [barebones](https://github.com/AjaniBilby/remix-worker-error/tree/barebones) so I can rollback later, interestingly if you uncomment out [these lines](https://github.com/AjaniBilby/remix-worker-error/blob/barebones/app/pdf/interface.server.ts#L66-L76) the entire server explodes attempting to rebind the already in use port.

Interestingly this isn't a silent error that always happens because with that line commented out the `worker.threadId`, keeps changing so the main thread keeps restarting or being spawned.

**Comment by AjaniBilby** [comment]
If you change this in the [`interface.server.ts`](https://github.com/AjaniBilby/remix-worker-error/blob/barebones/app/pdf/interface.server.ts):
```ts
process.on('uncaughtException', (error) => {
	console.log(`\x1b[95m[main-client]\x1b[0m \x1b[36mUncaught Exception\x1b[0m`, error);
});
```
To
```
process.on('uncaughtException', (error) => {
	console.log(`\x1b[95m[main-client]\x1b[0m \x1b[36mUncaught Exception\x1b[0m`, error);
	process.exit(1);
});
```
You don't get the rebind issue, nor does that log message ever get triggered so theoreticallly adding that line should have made no difference... this is mildly terrifying

**Comment by AjaniBilby** [comment]
I think I found something close to the true symptom, I restarted from the [`blue-stacks` example](https://github.com/remix-run/blues-stack), and I've got the workers successfully restarting on on reload as percieved in [`interface.server.ts`](https://github.com/AjaniBilby/blues-stack/blob/worker-example/app/worker/interface.server.ts), however somehow the routes themselves aren't accessing the new worker `interface.server.ts` when it's hot reloaded, which means they attempt to talk to a worker which has been destoryed and is no longer referenced in the singleton.

![hot-reload](https://github.com/remix-run/remix/assets/11359344/3af25024-7a36-4b51-a7c7-21d55f873417)

I'm not sure quite how one file get's one output from a singleton while another gets an outdated one, when they're both running of the same single access call...
But it does mean something really weird is happening with global and local state during hot reloads

**Comment by AjaniBilby** [comment]
I'm not sure directly how, but the fix in pr [blues-stack#222](https://github.com/remix-run/blues-stack) fixed the issue ([example using comlink](https://github.com/AjaniBilby/blues-stack/tree/84e1007d8fbc1f6f7e98e1051cb220c5892383ef))
